test/rule_definition_tools.jl

```diff
@@ -254,21 +254,54 @@ end
 
 
 module IsolatedModuleForTestingScoping
-    using Test
-    # need to make sure macros work in something that hasn't imported all exports
-    # all that matters is that the following don't error, since they will resolve at
-    # parse time
-    using ChainRulesCore: ChainRulesCore
+using ChainRulesCore: @scalar_rule, @non_differentiable
 
-    @testset "@non_differentiable" begin
-        # this is
-        # https://github.com/JuliaDiff/ChainRulesCore.jl/issues/317
-        fixed(x) = :abc
-        ChainRulesCore.@non_differentiable fixed(x)
-    end
+# ensure that functions, types etc. in module `ChainRulesCore` can't be resolved
+const ChainRulesCore = nothing
 
-    @testset "@scalar_rule" begin
-        my_id(x) = x
-        ChainRulesCore.@scalar_rule(my_id(x), 1.0)
+# this is
+# https://github.com/JuliaDiff/ChainRulesCore.jl/issues/317
+fixed(x) = :abc
+@non_differentiable fixed(x)
+
+# check name collision
+fixed_kwargs(x; kwargs...) = :abc
+@non_differentiable fixed_kwargs(kwargs)
+
+my_id(x) = x
+@scalar_rule(my_id(x), 1.0)
+
+module IsolatedSubmodule
```

Comment:
That all `frule` and `rrule`s can be called without errors - this would detect e.g. if the macro output would try to look up something in `ChainRulesCore` in the module the rules were defined in (i.e., in `IsolatedModuleForTestingScoping.ChainRulesCore`). To avoid having to import anything from ChainRulesCore in addition to `@scalar_rule` and `@non_differentiable`, these checks are performed in a separate submodule.

Reply:
Done :+1: